### PR TITLE
8333248: VectorGatherMaskFoldingTest.java failed when maximum vector bits is 64

### PR DESCRIPTION
**Pull request by zifeihan**
Hi, VectorGatherMaskFoldingTest.java Test fails when max vector bits is 64, when max vector bits is 64, LongVector.SPECIES_MAX.length() and DoubleVector.SPECIES_MAX.length() is 1. 

We can reproduce this problem in two ways:
1. We can use riscv without rvv1.0 board to reproduce this problem
2. Run VectorGatherMaskFoldingTest.java on aarch64 client mode without `-XX:+IncrementalInlineForceCleanup` Option,  the `-XX:+IncrementalInlineForceCleanup` is C2 Option, so we need to remove this Option from the VectorGatherMaskFoldingTest.main method. error message:
```
Base Test: @Test testDoubleVectorStoreLoadMaskedVector:
compiler.lib.ir_framework.shared.TestRunException: There was an error while invoking @Test method public static void compiler.vectorapi.VectorGatherMaskFoldingTest.testDoubleVectorStoreLoadMaskedVector(). Target: null. Arguments: <void>
	at compiler.lib.ir_framework.test.BaseTest.invokeTestMethod(BaseTest.java:84)
	at compiler.lib.ir_framework.test.BaseTest.invokeTest(BaseTest.java:71)
	at compiler.lib.ir_framework.test.AbstractTest.run(AbstractTest.java:98)
	at compiler.lib.ir_framework.test.TestVM.runTests(TestVM.java:861)
	at compiler.lib.ir_framework.test.TestVM.start(TestVM.java:252)
	at compiler.lib.ir_framework.test.TestVM.main(TestVM.java:165)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at compiler.lib.ir_framework.test.BaseTest.invokeTestMethod(BaseTest.java:80)
	... 5 more
Caused by: java.lang.RuntimeException: assertNotEquals: expected [1.0] to not equal [1.0]
	at jdk.test.lib.Asserts.fail(Asserts.java:691)
	at jdk.test.lib.Asserts.assertNotEquals(Asserts.java:451)
	at jdk.test.lib.Asserts.assertNotEquals(Asserts.java:435)
	at compiler.vectorapi.VectorGatherMaskFoldingTest.testDoubleVectorStoreLoadMaskedVector(VectorGatherMaskFoldingTest.java:1089)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 7 more
```

For example, the following method will be failed:
```
    private static final VectorSpecies<Long> L_SPECIES = LongVector.SPECIES_MAX;
    private static final VectorSpecies<Double> D_SPECIES = DoubleVector.SPECIES_MAX;
    ...
    @Test
    @IR(counts = { IRNode.STORE_VECTOR_MASKED, ">= 1", IRNode.LOAD_VECTOR_MASKED, ">= 1" }, applyIfCPUFeatureOr = {"avx512", "true", "sve", "true"})
    public static void testDoubleVectorStoreLoadMaskedVector() {
        double[] res = new double[D_SPECIES.length()];
        doubleVector.intoArray(res, 0, doubleVectorMask);
        DoubleVector res2 = DoubleVector.fromArray(D_SPECIES, res, 0, doubleVectorMask);
        Asserts.assertNotEquals(res2, doubleVector);
    }
```

in this `testDoubleVectorStoreLoadMaskedVector` test case, the doubleVector data is:[1.0], doubleVectorMask:[true], res2 is:[1.0] So here `Asserts.assertNotEquals(res2, doubleVector);` will assert Error.

By the way, LongVector.SPECIES_MAX/ DoubleVector.SPECIES_MAX is initialized with a call to VectorShape.getMaxVectorBitSize. the aarch64 client jvm mode, VectorShape.getMaxVectorBitSize will return the default 64 bit, and If any CPU does not support vectors like riscv without rvv1.0, the default value of 64 is returned.
```
    /**
     * Returns the maximum vector bit size for a given element type.
     *
     * @param etype the element type.
     * @return the maximum vector bit.
     */
     /*package-private*/
    static int getMaxVectorBitSize(Class<?> etype) {
        // VectorSupport.getMaxLaneCount may return -1 if C2 is not enabled,
        // or a value smaller than the S_64_BIT.vectorBitSize / elementSizeInBits if MaxVectorSize < 16
        // If so default to S_64_BIT
        int maxLaneCount = VectorSupport.getMaxLaneCount(etype);
        int elementSizeInBits = LaneType.of(etype).elementSize;
        return Math.max(maxLaneCount * elementSizeInBits, S_64_BIT.vectorBitSize);
    }
```

The fix:
1. make sure that masks are not the same for lengths of 1 (for long and double vectors): something like this `longMask[i] = L_SPECIES.length() > 1 && i % 2 == 0`;
2. For long and double indices this cannot be done with length 1. So change the asserts to add a check for the length, e.g. `Asserts.assertFalse(L_SPECIES.length() != 1 && res.equals(res2));`

### Testing
- [x] Run VectorGatherMaskFoldingTest.java on Banana Pi BPI-F3 board (has RVV1.0)
- [x] Run VectorGatherMaskFoldingTest.java on SOPHON SG2042 (without RVV1.0)
- [x] Run VectorGatherMaskFoldingTest.java on aarch64 server mode with neon
- [x] Run VectorGatherMaskFoldingTest.java on aarch64 client mode without `-XX:+IncrementalInlineForceCleanup` Option

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333248](https://bugs.openjdk.org/browse/JDK-8333248): VectorGatherMaskFoldingTest.java failed when maximum vector bits is 64 (**Bug** - P4)(⚠️ The fixVersion in this issue is [23] but the fixVersion in .jcheck/conf is 24, a new backport will be created when this pr is integrated.)

### Reviewers
 * [Damon Fenacci](https://openjdk.org/census#dfenacci) (@dafedafe - Committer)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19473/head:pull/19473` \
`$ git checkout pull/19473`

Update a local copy of the PR: \
`$ git checkout pull/19473` \
`$ git pull https://git.openjdk.org/jdk.git pull/19473/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19473`

View PR using the GUI difftool: \
`$ git pr show -t 19473`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19473.diff">https://git.openjdk.org/jdk/pull/19473.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19473#issuecomment-2139273277)

### CODE REVIEW

**Comment by zifeihan** [comment]
@dafedafe Maybe you can take a look? Thanks.

**Comment by dafedafe** [comment]
> @dafedafe Hi Damon, Maybe you can take a look at this small change? Thanks.

@zifeihan you're right that the case with a vector length of 64 bits is not covered correctly.

I wouldn't fix it using different species depending on the length though. The Java Vector API explicitly says that not all shapes are supported by all platforms, and one should choose shape-agnostic code. Also, the IR framework uses the maximum vector size possible (if not explicitly set in the rules).
I'd rather make sure that masks are not the same for lengths of 1 (for long and double vectors): something like this `longMask[i] = L_SPECIES.length() > 1 && i % 2 == 0;` would probably do.
For long and double indices this cannot be done with length 1. So an option would be to change the asserts to add a check for the length, e.g. `Asserts.assertFalse(L_SPECIES.length() != 1 && res.equals(res2));`.
There is always the alternative of not running long and double vector tests with max vector lengths of 64.

Regarding `-XX:+IncrementalInlineForceCleanup`: this has been added because of the issue described in [JDK-8302459](https://bugs.openjdk.org/browse/JDK-8302459) that sometimes prevents C2 to properly vectorise (it should be removed when this issue is fixed). Having it there shouldn't really make any test fail though. If this is the case with `aarch64` it might mean that there is another issue somewhere else.

**Comment by zifeihan** [comment]
> > @dafedafe Hi Damon, Maybe you can take a look at this small change? Thanks.
> 
> @zifeihan you're right that the case with a vector length of 64 bits is not covered correctly.
> 
> I wouldn't fix it using different species depending on the length though. The Java Vector API explicitly says that not all shapes are supported by all platforms, and one should choose shape-agnostic code. Also, the IR framework uses the maximum vector size possible (if not explicitly set in the rules). I'd rather make sure that masks are not the same for lengths of 1 (for long and double vectors): something like this `longMask[i] = L_SPECIES.length() > 1 && i % 2 == 0;` would probably do. For long and double indices this cannot be done with length 1. So an option would be to change the asserts to add a check for the length, e.g. `Asserts.assertFalse(L_SPECIES.length() != 1 && res.equals(res2));`. There is always the alternative of not running long and double vector tests with max vector lengths of 64.
> 

Hi Damon, I have done the fix. Can you talk a look again? Thanks

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 1]
```diff
@@ -160,23 +160,23 @@ public class VectorGatherMaskFoldingTest {
     public static void testTwoLongVectorLoadGatherNotEqualArray() {
         LongVector res = LongVector.fromArray(L_SPECIES, longArray, 0, longIndices, 0);
         LongVector res2 = LongVector.fromArray(L_SPECIES, longArray2, 0, longIndices, 0);
-        Asserts.assertNotEquals(res, res2);
+        Asserts.assertFalse(L_SPECIES.length() != 1 && res.equals(res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:163]: I think you don't need this change (`longArray` and `longArray2` are different even with length 1).

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 2]
```diff
@@ -224,15 +224,15 @@ public static void testTwoLongVectorLoadGatherMaskedEquals() {
     public static void testTwoLongVectorLoadGatherMaskedNotEqualMask() {
         LongVector res = LongVector.fromArray(L_SPECIES, longArray, 0, longIndices, 0, longVectorMask);
         LongVector res2 = LongVector.fromArray(L_SPECIES, longArray, 0, longIndices, 0, longVectorMask2);
-        Asserts.assertNotEquals(res, res2);
+        Asserts.assertFalse(L_SPECIES.length() != 1 && res.equals(res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:227]: You don't need this change either (`longVectorMask` and `longVectorMask` are different even with length 1).

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 3]
```diff
@@ -326,23 +326,23 @@ public static void testTwoIntVectorLoadGatherMaskedNotEqualIndices() {
     public static void testTwoDoubleVectorLoadGatherNotEqualArray() {
         DoubleVector res = DoubleVector.fromArray(D_SPECIES, doubleArray, 0, doubleIndices, 0);
         DoubleVector res2 = DoubleVector.fromArray(D_SPECIES, doubleArray2, 0, doubleIndices, 0);
-        Asserts.assertNotEquals(res, res2);
+        Asserts.assertFalse(D_SPECIES.length() != 1 && res.equals(res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:329]: Same here.

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 4]
```diff
@@ -390,15 +390,15 @@ public static void testTwoDoubleVectorLoadGatherMaskedEquals() {
     public static void testTwoDoubleVectorLoadGatherMaskedNotEqualMask() {
         DoubleVector res = DoubleVector.fromArray(D_SPECIES, doubleArray, 0, doubleIndices, 0, doubleVectorMask);
         DoubleVector res2 = DoubleVector.fromArray(D_SPECIES, doubleArray, 0, doubleIndices, 0, doubleVectorMask2);
-        Asserts.assertNotEquals(res, res2);
+        Asserts.assertFalse(D_SPECIES.length() != 1 && res.equals(res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:393]: Same here.

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 5]
```diff
@@ -576,7 +576,7 @@ public static void testTwoLongVectorStoreScatterMaskedNotEqualMask() {
         long[] res2 = new long[L_SPECIES.length()];
         longVector.intoArray(res, 0, longIndices, 0, longVectorMask);
         longVector.intoArray(res2, 0, longIndices, 0, longVectorMask2);
-        Asserts.assertFalse(Arrays.equals(res, res2));
+        Asserts.assertFalse(L_SPECIES.length() != 1 && Arrays.equals(res, res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:579]: Same here.

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 6]
```diff
@@ -702,7 +702,7 @@ public static void testTwoDoubleVectorStoreScatterNotEqualVector() {
         double[] res2 = new double[D_SPECIES.length()];
         doubleVector.intoArray(res, 0, doubleIndices, 0);
         doubleVector2.intoArray(res2, 0, doubleIndices, 0);
-        Asserts.assertFalse(Arrays.equals(res, res2));
+        Asserts.assertFalse(D_SPECIES.length() != 1 && Arrays.equals(res, res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:705]: Same here.

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java [hunk 7]
```diff
@@ -782,7 +782,7 @@ public static void testTwoDoubleVectorStoreScatterMaskedNotEqualMask() {
         double[] res2 = new double[D_SPECIES.length()];
         doubleVector.intoArray(res, 0, doubleIndices, 0, doubleVectorMask);
         doubleVector.intoArray(res2, 0, doubleIndices, 0, doubleVectorMask2);
-        Asserts.assertFalse(Arrays.equals(res, res2));
+        Asserts.assertFalse(D_SPECIES.length() != 1 && Arrays.equals(res, res2));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:785]: Same here.

**Review by dafedafe — COMMENTED** [review]
On the other hand you might need to add the check here:
https://github.com/openjdk/jdk/blob/1235a453eef4a838fe07009b0c5e8a962b527bb4/test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java#L966
and here:
https://github.com/openjdk/jdk/blob/1235a453eef4a838fe07009b0c5e8a962b527bb4/test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java#L1098

**Comment by zifeihan** [comment]
> On the other hand you might need to add the check here:
> 
> https://github.com/openjdk/jdk/blob/1235a453eef4a838fe07009b0c5e8a962b527bb4/test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java#L966
> 
> 
> and here:
> https://github.com/openjdk/jdk/blob/1235a453eef4a838fe07009b0c5e8a962b527bb4/test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java#L1098

fixed.

File: test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java (hunk 7 above)
Comment (zifeihan) [reply test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java]: I've fixed all of the above comments.

**Comment by dafedafe** [comment]
It looks good now, thanks!
Have you tried running the test again with a 64-bit vector platform?
I'd also update the description of this PR.

**Comment by zifeihan** [comment]
> It looks good now, thanks! Have you tried running the test again with a 64-bit vector platform? I'd also update the description of this PR.

Yes. This case still pass on 64-bit vector platforms, including aarch64 client/server build and linux-riscv64 (with/without vector extension).

**Comment by zifeihan** [comment]
@dafedafe : Hi, Do you have any other suggestions? Thanks

**Comment by dafedafe** [comment]
@zifeihan it might be good to update the description of this PR as well. Thanks!

**Comment by zifeihan** [comment]
> @zifeihan it might be good to update the description of this PR as well. Thanks!

Hi, I've updated the description of this PR. 
Do you have any other suggestions? Thanks

**Comment by dafedafe** [comment]
In the description you mention
>Run VectorGatherMaskFoldingTest.java on aarch64 client mode without -XX:+IncrementalInlineForceCleanup Option, the -XX:+IncrementalInlineForceCleanup is C2 Option, so we need to remove this Option from the VectorGatherMaskFoldingTest.main method

I'm wondering why this happens: the test needs C2, so a C2 option should definitely be OK. Additionally what this option does is cleaning up more often when doing late inlining (it shouldn't be detrimental). Did you find out why you get an error? What kind of error is it?

**Comment by zifeihan** [comment]
> I'm wondering why this happens: the test needs C2, so a C2 option should definitely be OK. Additionally what this option does is cleaning up more often when doing late inlining (it shouldn't be detrimental). Did you find out why you get an error? What kind of error is it?

Sorry, I didn't express myself clearly, here in C2 there is no problem, if use client vm to test it, client vm doesn't have C2,so we need to remove this option, because this option is defined by C2, in client vm can't find this jvm option.

**Comment by dafedafe** [comment]
👍 that's clear now thanks! To avoid this issue you can insert the `-XX:+IgnoreUnrecognizedVMOptions` flag before `-XX:+IncrementalInlineForceCleanup`.

**Comment by zifeihan** [comment]
> 👍 that's clear now thanks! To avoid this issue you can insert the `-XX:+IgnoreUnrecognizedVMOptions` flag before `-XX:+IncrementalInlineForceCleanup`.

Thanks for the suggestion, Fixed.

**Review by RealFYang — APPROVED** [review]
LGTM.

**Comment by zifeihan** [comment]
@dafedafe @RealFYang : Thanks for the review.
/integrate

**Comment by robehn** [comment]
This doesn't seem to made it to jdk23 AFAICT ?
Can we backport it?

**Comment by robehn** [comment]
> /backport jdk23u

Thank you!

**Pull request closed without merging**